src/swap/do_action.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{swap::db, SwapId};
+use comit::Timestamp;
+
+#[async_trait::async_trait]
+pub trait Do<E>
+where
+    Self: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+    async fn r#do(
+        &self,
+        beta_expiry: Timestamp,
+        execution_args: <Self as Execute<E>>::Args,
+    ) -> anyhow::Result<Next<E>> {
+        if let Some(event) = self.check_memory().await? {
+            return Ok(Next::Continue(event));
+        }
+
+        if self.should_abort(beta_expiry).await? {
+            return Ok(Next::Abort);
+        }
+
+        let event = Execute::<E>::execute(self, execution_args).await?;
+        self.remember(event.clone()).await?;
+
+        Ok(Next::Continue(event))
+    }
+}
+
+#[async_trait::async_trait]
+impl<E, A> Do<E> for A
+where
+    A: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+}
+
+#[async_trait::async_trait]
+pub trait CheckMemory<E> {
+    async fn check_memory(&self) -> anyhow::Result<Option<E>>;
+}
+
+#[async_trait::async_trait]
+impl<E, A> CheckMemory<E> for A
+where
+    A: db::Load<E> + std::ops::Deref<Target = SwapId>,
+    E: 'static,
+{
+    async fn check_memory(&self) -> anyhow::Result<Option<E>> {
+        self.load(**self).await
+    }
+}
+
+#[async_trait::async_trait]
+pub trait ShouldAbort {
+    async fn should_abort(&self, beta_expiry: Timestamp) -> anyhow::Result<bool>;
+}
+
+#[async_trait::async_trait]
+impl<A> ShouldAbort for A
+where
+    A: BetaLedgerTime + Sync,
+{
+    async fn should_abort(&self, beta_expiry: Timestamp) -> anyhow::Result<bool> {
+        let beta_ledger_time = self.beta_ledger_time().await?;
+
+        Ok(beta_expiry <= beta_ledger_time)
+    }
+}
+
+#[async_trait::async_trait]
+pub trait Execute<E> {
+    type Args;
+    async fn execute(&self, args: Self::Args) -> anyhow::Result<E>;
+}
+
+#[async_trait::async_trait]
+pub trait Remember<E> {
```

Comment:
Can you please comment the expected functionality brought by the various traits `Remember`, `Execute`, `CheckMemory`, `Do`?

Reply:
You thumbed up but did not do it?

Reply:
I _will_ do it.